### ISSUE TITLE
`Error x509: certificate relies on legacy Common Name field, use SANs instead`

### ISSUE THREAD

**Issue by chris001** [virtualmin/virtualmin-gpl#462]
EDIT:  Current version of many google and cloudflare services written in Go language, will fail with old CN certs, they require update new SAN certs,

Can Virtualmin update the self-signed cert generating code?  The legacy `CN (Common Name)` now fails.  Must use the current stable `SAN (Subject Alternative Name)`

**Comment by chris001** [comment]
This error is because the Common Name has been deprecated!  Solution below!
Reference: https://github.com/rancher/rke2/issues/775
RFC2818: https://tools.ietf.org/html/rfc2818#section-3.1

>    If a `subjectAltName` extension of type `dNSName` is present, that MUST be used as the identity. Otherwise, the (most specific) `Common Name` field in the `Subject` field of the certificate MUST be used. Although the use of the `Common Name` is existing practice, it is deprecated and Certification Authorities are encouraged to use the `Name` instead.

This can be fixed by generating the certificate with the procedure given below:
Reference: https://stackoverflow.com/questions/64814173/how-do-i-use-sans-with-openssl-instead-of-common-name

**SOLUTION**
3 Steps to generate this self signed cert with the RFC2818 -required `SAN` (`Subject Alternative Name`) field, which will work with modern standards-compliant web services e.g. CloudFlare and Google:
1. You must have to have a CA (local CA works too)
```
openssl genrsa -out ca.key 2048
openssl req -new -x509 -days 365 -key ca.key -subj "/C=IN/ST=KA/L=BL/O=MyOrg, Inc./CN=MyOrg Root CA" -out ca.crt
```
2. Create a server CSR with 'localhost' in CN:
`openssl req -newkey rsa:2048 -nodes -keyout server.key -subj "/C=IN/ST=KA/L=BL/O=MyOrg, Inc./CN=localhost" -out server.csr`

3. Finally, sign the server cert by CA and pass the `subjectAltName` when you signing the server certificate.
`openssl x509 -req -extfile <(printf "subjectAltName=DNS:localhost") -days 365 -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt`

**Comment by jcameron** [comment]
Are the CSRs generated by Virtualmin using the wrong certificate fields?

**Comment by chris001** [comment]
> Are the CSRs generated by Virtualmin using the wrong certificate fields?

Without looking at the Virtualmin code yet, Yes.   Based on the failure / error message, received from the Go `https` web server (Cloudflare Tunnel for self hosting Virtualmin with no IP address behind double NAT) used by Cloudflare web services (Google Chrome also requires this SAN field).

**Comment by marclaporte** [comment]
@chris001 discovered this while working on https://forum.virtualmin.com/t/investigating-for-a-solution-to-permit-self-hosting-without-a-fixed-ip-address-such-as-pagekite/116779/28

**Comment by jcameron** [comment]
So, the current Webmin/Virtualmin code will set both the CN field to the first domain name, and subjectAltName to a list of all the domain names in the cert. So it should be compatible with all use cases..

**Comment by chris001** [comment]
I'll have to check the virtualmin generated self signed certs closer then and see if the info is in the right place and in the expected format.  When you start the tunnel, cloudflare tries to contact virtualmin with `https` on port `10000` which is using the self generated self signed cert.  The CF error message could be inaccurate, in which case the cause would be something else.  Will check.

**Comment by chris001** [comment]
> So, the current Webmin/Virtualmin code will set both the CN field to the first domain name, and subjectAltName to a list of all the domain names in the cert. So it should be compatible with all use cases..

Does the self-signed cert include `localhost` as one of the names in the SAN field?  I'm sure it doesn't.  If no, can we update the self signed cert generation code to include `localhost` by default?  95% sure, this is why GoLang TLS client web service is failing the cert check on our self-signed cert.  We're trying to serve `https://localhost:10000` so the name `localhost` needs to be in the cert's `SAN` field.  (The working example _also_ includes `localhost` in the `CN` field however that `CN` field should be unimportant for the cert check.)

**Comment by iliajie** [comment]
Interesting, although it is somewhat surprising to hear that self-signed certificate cares about CN accuracy.

**Comment by chris001** [comment]
I agree, Go is strict and up to date with these RFCs.  Big discussion topic.  It's the GoLang built-in cert checking, running on CloudFlare, in this case the CloudFlare Tunnel trying to connect to the localhost origin web server running the other end of the cloudflare tunnel, that is giving this error about the cert.

**Comment by chris001** [comment]
I found this:
```
$ echo | openssl s_client -connect redhat.com:443 2>/dev/null | openssl x509 -noout -ext subjectAltName
X509v3 Subject Alternative Name:
    DNS:redhat.com, DNS:www.redhat.com
$ echo | openssl s_client -connect localhost:10000 2>/dev/null | openssl x509 -noout -ext subjectAltName
No extensions in certificate
```
Looks like Virtualmin/Webmin is not using the `subjectAltName` extension in the self-signed cert.  
This would explain why GoLang is throwing this error and refuses to connect the Tunnel to virtualmin on `localhost` port `10000`.

**Comment by iliajie** [comment]
Can you confirm that using `subjectAltName` in self-signed certificate for Webmin makes connections to Webmin's _miniserv_ work with GoLang?

**Comment by chris001** [comment]
Getting closer!  With the addition of `subjectAltName` on a new self-signed cert, the connection error to `miniserv` is now, `x509: certificate signed by unknown authority`.  Circular problem now because the new Virtualmin installed on `localhost` does not have a public hostname and will not necessarily have one because it will often be located behind double NAT, it's a new user trying to allow public access to their webmin with this temporary CF `https` `URL`.  Time to see if the CF tunnel would skip verifying the self signed `subjectAltName` TLS cert's `CA`. Unless `webmin` can provide to CF its own `CA` cert used to sign the `https` cert, and ask CF to trust it for this temporary quick tunnel connection?

**Comment by chris001** [comment]
To answer your question, Yes, the `subjectAltName` must be added to the cert, before GoLang will attempt to verify the cert (or skip verification on user request) and connect to Webmin `miniserv`.

**Comment by chris001** [comment]
Got it!  CF Quick Tunnel connects to new install Virtualmin on `localhost` with no working domain name!  CF only requires a properly made recent RFC compliant self signed cert with the `subjetcAltName` in it!  Skips verifying the `CA` and connects!
![cloudflare quick tunnel works virtualmin](https://user-images.githubusercontent.com/259416/194176705-1bc6c30d-6944-47ef-bb0e-43b22fc7c7cd.png)

**Comment by chris001** [comment]
I could add a feature option to the quick tunnel script where only upon user request, it'll generate and install in Virtualmin the RFC compliant self signed cert with `subjectAltName`, so that the CF Quick Tunnel will allow the cert and connect.

**Comment by jcameron** [comment]
Ah, so the real issue seems to be that the self-signed cert Webmin creates for itself at install time doesn't use subjectAltName? That's fixable, along with including localhost if we want to.

**Comment by chris001** [comment]
Yes.  I've found that, the `localhost` bit is somewhat optional, when it's skipping the cert verification, it doesn't care that the `https` URL `hostname` doesn't match any `hostname` on the self signed cert, however, not 100% sure of this yet, haven't tried it in every combination.

**Comment by jcameron** [comment]
Ok, I'll add this in the next release..

**Comment by chris001** [comment]
In case anyone wants to try and get a quick `https` URL to your new install of virtualmin on your `localhost` linux PC computer, without having your own IP address or domain name pointed to your computer because you're on a typical internet service provider.... Here's instructions to install a working self signed cert/key, and get your temporary `https` URL to your login page!

1. Download, or open in a text viewer, these 2 text files on your computer: [selfsigned.crt.txt](https://github.com/virtualmin/virtualmin-gpl/files/9728893/selfsigned.crt.txt)
[selfsigned.key.txt](https://github.com/virtualmin/virtualmin-gpl/files/9728896/selfsigned.key.txt)
2. In your browser login to your `localhost` Virtualmin
3. Click `Virtualmin` > `Server Configuration` > `SSL Certificate` > `Update Certificate And Key`, 
4. Click `Signed SSL Cert` select `Pasted Text`, open the `selfsigned.crt.txt` file, copy all the contents to your clipboard (Ctrl-A Ctrl-C), click the text box, paste it in (Ctrl-V).
5. Under `Matching Private Key`, select `Pasted Text`, open the `selfsigned.key.txt` file you downloaded from this post, copy all the contents to your clipboard (Ctrl-A Ctrl-C), click the text box, paste it in (Ctrl-V).
6. Under `Private key password` select `None needed`.
7. Click `Install now`.
8. Restart Virtualmin again manually. In your browser, in virtualmin, click the icon ">_" (Command shell or Alt-K).  Type the command `sudo systemctl restart webmin` and hit enter.  (Why? Because of a possible bug, Virtualmin didn't restart and use this new cert and key, so you had to run that command.)
9. Go to your usual `ssh` terminal, logged into your `localhost` Virtualmin linux server. 
10. Copy paste this one line command into your terminal and hit enter: `wget -q -O - https://github.com/chris001/CGNAThome/raw/main/start-tunnel.sh | bash`
11. Carefully watch the text on screen, for your random temporary `https` `trycloudflare` URL.  
12. Highlight that `https` `trycloudflare` URL, to auto copy it from the terminal app, to your clipboard.
13. Go to browser and make a new tab (Ctrl-T). Paste that URL in address bar (Ctrl-V), hit enter, it should load your Virtualmin login page!
14. If you want, Share your temporary `https` URL to anyone (co-worker, tech support, client, engineer, associate). With this URL they'll gain access to your Virtualmin login page.  They can be located anywhere on the global internet!
15. How long will this URL work?  Your temporary `https` URL to your localhost virtualmin, will continue to work UNTIL you press `Ctrl-C` in your `ssh` terminal where you ran the start-tunnel command (in step 10 above).
16. Enjoy and comment back.

**Comment by iliajie** [comment]
@chris001 Chris, have a look please at this comment and a correspondent commit.

https://github.com/webmin/webmin/commit/57537ef282d9e8ec1bae786b3147701ee236beff#commitcomment-86349174

If you generate a certificate the way we do it now, does it work for your app?

**Comment by iliajie** [comment]
This is the final and the correct patch:

https://github.com/webmin/webmin/commit/ea0e66cff5ed26a37dbc18444f073a6eba68b03c

**Comment by chris001** [comment]
Hi Ilia, I'm a bit late, however, these commands from [forum thread](https://forum.virtualmin.com/t/investigating-for-a-solution-to-permit-self-hosting-without-a-fixed-ip-address-such-as-pagekite/116779/35?u=chris_c) are working:
```
$ openssl genrsa -out ca.key 2048
$ openssl req -new -x509 -days 365 -key ca.key -subj "/C=IN/ST=KA/L=BL/O=Chris001, Inc./CN=Chris001 Root CA" -out ca.crt
$ openssl req -newkey rsa:2048 -nodes -keyout server.key -subj "/C=IN/ST=KA/L=BL/O=Chris001, Inc./CN=localhost" -out server.csr
............+..+.+.........+...+.....+...+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.+..............+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*..+.....+...+....+.........+............+.....+...+....+..+....+.....+.+.........+.....+.......+............+..+...............+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
.....+.............+..+...+............+...+.+.....+.+........+.......+..+.......+........+...+...+.+...+...+.....+........................+...+.............+..+...+....+......+..+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*..+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++*.....+....................+...+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
-----
$ openssl x509 -req -extfile <(printf "subjectAltName=DNS:localhost") -days 365 -in server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out server.crt
Certificate request self-signature ok
subject=C = IN, ST = KA, L = BL, O = "Chris001, Inc.", CN = localhost
```

**Comment by iliajie** [comment]
No, please try this variant instead:

```
openssl req -newkey rsa:2048 -x509 -nodes -out ssl.cert -keyout ssl.key -days 1825 -sha256 -subj "/CN=machine1.host.name/C=US/L=Santa Clara" -addext subjectAltName=DNS:machine1.host.name,DNS:localhost -addext extendedKeyUsage=serverAuth
```

**Comment by chris001** [comment]
Ilia, Yes, ran your variant, installed cert/key in Virtualmin, it works, CF connects the tunnel with the cert/key generated by your command, even though it's to a different `locahost` configured with the persistent tunnel and user own domain name, not the random temporary `.trycloudflare.com` subdomain.  Your one line command is brilliant compared with the example 4 lines command!

**Comment by iliaross** [comment]
Reopened, as this was closed by an automated cleanup run.

⚠️ If the issue is resolved, please take a moment to close it to help keep the project organized.
